File: src/backend/php/app/Services/UserService.php
<?php

namespace App\Services;

use App\Exceptions\InvalidPasswordException;
use App\Exceptions\InvalidPasswordResetTokenException;
use App\Exceptions\UserConflictException;
use App\Http\Requests\ChangePasswordRequest;
use App\Http\Requests\ConfirmPasswordResetRequest;
use App\Http\Requests\CreateTokenRequest;
use App\Http\Requests\CreateUserRequest;
use App\Http\Requests\ResetPasswordRequest;
use App\Http\Requests\SignUpRequest;
use App\Http\Requests\UpdateProfileRequest;
use App\Http\Requests\UpdateUserRequest;
use App\Models\User;
use App\Repositories\Contracts\UserRepositoryInterface;
use App\Services\Contracts\UserServiceInterface;
use Firebase\JWT\JWT;
use Illuminate\Contracts\Encryption\DecryptException;
use Illuminate\Database\Eloquent\ModelNotFoundException;
use Illuminate\Pagination\LengthAwarePaginator;
use Illuminate\Support\Facades\Crypt;
use Illuminate\Support\Facades\Hash;
use JsonException;

class UserService implements UserServiceInterface
{
    private ?string $dummyPasswordHash = null;

    public function __construct(
        private readonly UserRepositoryInterface $repository
    ) {}

    public function getAll(int $page = 1, int $perPage = 20): LengthAwarePaginator
    {
        return $this->repository->paginate(max(1, $page), min(100, max(1, $perPage)));
    }

    public function getById(int $id): User
    {
        $user = $this->repository->findById($id);
        if (! $user instanceof User) {
            throw new ModelNotFoundException("User {$id} not found.");
        }

        return $user;
    }

    public function create(CreateUserRequest $request): User
    {
        $data = $request->validated();
        $username = trim($data['username']);
        $email = $this->normalizeEmail($data['email']);
        $this->ensureUnique($username, $email);

        $user = $this->repository->create([
            'username' => $username,
            'email' => $email,
            'password_hash' => Hash::make($data['password']),
            'is_active' => $data['is_active'] ?? true,
        ]);

        /** @var User $user */
        return $user;
    }

    public function update(int $id, UpdateUserRequest $request): User
    {
        $user = $this->getById($id);
        $data = $request->validated();
        $username = array_key_exists('username', $data) ? trim($data['username']) : $user->username;
        $email = array_key_exists('email', $data) ? $this->normalizeEmail($data['email']) : $user->email;
        $this->ensureUnique($username, $email, $id);

        $changes = ['username' => $username, 'email' => $email, 'updated_at' => now()];
        if (array_key_exists('password', $data)) {
            $changes['password_hash'] = Hash::make($data['password']);
        }

        /** @var User $updated */
        $updated = $this->repository->update($user, $changes);

        return $updated;
    }

    public function setActive(int $id, bool $isActive): User
    {
        $user = $this->getById($id);
        /** @var User $updated */
        $updated = $this->repository->update($user, ['is_active' => $isActive, 'updated_at' => now()]);

        return $updated;
    }

    public function signup(SignUpRequest $request): User
    {
        $data = $request->validated();
        $username = trim($data['username']);
        $email = $this->normalizeEmail($data['email']);
        $this->ensureUnique($username, $email);

        /** @var User $user */
        $user = $this->repository->create([
            'username' => $username,
            'email' => $email,
            'password_hash' => Hash::make($data['password']),
            'is_active' => true,
        ]);

        return $user;
    }

    public function getProfile(int $userId): User
    {
        return $this->getById($userId);
    }

    public function updateProfile(int $userId, UpdateProfileRequest $request): User
    {
        $user = $this->getById($userId);
        $data = $request->validated();
        $username = array_key_exists('username', $data) ? trim($data['username']) : $user->username;
        $email = array_key_exists('email', $data) ? $this->normalizeEmail($data['email']) : $user->email;
        $this->ensureUnique($username, $email, $userId);

        /** @var User $updated */
        $updated = $this->repository->update($user, [
            'username' => $username,
            'email' => $email,
            'updated_at' => now(),
        ]);

        return $updated;
    }

    public function changePassword(int $userId, ChangePasswordRequest $request): void
    {
        $user = $this->getById($userId);
        if (! $user->is_active) {
            throw new InvalidPasswordException('The user account is inactive.');
        }
        if (! Hash::check($request->validated('current_password'), $user->password_hash)) {
            throw new InvalidPasswordException('The current password is incorrect.');
        }

        $this->repository->update($user, [
            'password_hash' => Hash::make($request->validated('new_password')),
            'updated_at' => now(),
        ]);
    }

    public function requestPasswordReset(ResetPasswordRequest $request): void
    {
        $user = $this->repository->findByEmail($this->normalizeEmail($request->validated('email')));
        if ($user === null || ! $user->is_active) {
            return;
        }

        $lifetime = max(1, (int) config('users.password_reset_lifetime_minutes', 60));
        $token = Crypt::encryptString(json_encode([
            'user_id' => $user->id,
            'expires_at' => now()->addMinutes($lifetime)->getTimestamp(),
            'password' => hash('sha256', $user->password_hash),
        ], JSON_THROW_ON_ERROR));
        $baseUrl = (string) config('users.password_reset_confirmation_url', '/reset-password');
        $separator = str_contains($baseUrl, '?') ? '&' : '?';
        $url = $baseUrl.$separator.'token='.rawurlencode($token);

        $this->repository->createEmailLog([
            'recipient' => $user->email,
            'subject' => 'Reset your Todo API password',
            'body' => "Use this link to reset your password: {$url}\n\nThis link expires in {$lifetime} minutes.",
            'status' => 'pending',
        ]);
    }

    public function confirmPasswordReset(ConfirmPasswordResetRequest $request): void
    {
        try {
            $payload = json_decode(Crypt::decryptString($request->validated('token')), true, flags: JSON_THROW_ON_ERROR);
            $user = isset($payload['user_id']) ? $this->repository->findById((int) $payload['user_id']) : null;
            $valid = $user instanceof User
                && $user->is_active
                && isset($payload['expires_at'], $payload['password'])
                && (int) $payload['expires_at'] >= now()->getTimestamp()
                && hash_equals((string) $payload['password'], hash('sha256', $user->password_hash));
            if (! $valid) {
                throw new InvalidPasswordResetTokenException('The password reset token is invalid or expired.');
            }
        } catch (DecryptException|JsonException $exception) {
            throw new InvalidPasswordResetTokenException('The password reset token is invalid or expired.', previous: $exception);
        }

        $this->repository->update($user, [
            'password_hash' => Hash::make($request->validated('new_password')),
            'updated_at' => now(),
        ]);
    }

    public function createToken(CreateTokenRequest $request): ?array
    {
        $email = strtolower(trim((string) $request->validated('email')));
        $password = (string) $request->validated('password');
        $user = $this->repository->findByEmail($email);
        $this->dummyPasswordHash ??= Hash::make('not-a-real-password');
        $passwordValid = Hash::check($password, $user?->password_hash ?? $this->dummyPasswordHash);

        if ($user === null || ! $passwordValid || ! $user->is_active) {
            return null;
        }

        $issuedAt = time();
        $expiresIn = max(1, (int) config('users.jwt_token_lifetime_minutes', 60)) * 60;
        $token = JWT::encode([
            'sub' => (string) $user->getKey(),
            'iat' => $issuedAt,
            'exp' => $issuedAt + $expiresIn,
        ], (string) config('users.jwt_secret'), 'HS256');

        return [
            'access_token' => $token,
            'token_type' => 'Bearer',
            'expires_in' => $expiresIn,
        ];
    }

    private function ensureUnique(string $username, string $email, ?int $excludingId = null): void
    {
        if ($this->repository->usernameExists($username, $excludingId)) {
            throw new UserConflictException('Username is already in use.');
        }
        if ($this->repository->emailExists($email, $excludingId)) {
            throw new UserConflictException('Email is already in use.');
        }
    }

    private function normalizeEmail(string $email): string
    {
        return strtolower(trim($email));
    }
}

File: src/backend/dotnet/TodoApi/Services/UserService.cs
using System.Text.Json;
using Microsoft.AspNetCore.DataProtection;
using Microsoft.AspNetCore.Identity;
using TodoApi.Data;
using TodoApi.DTOs;
using TodoApi.Repositories;
using TodoShared.Models;

namespace TodoApi.Services;

public class UserService(
    IUserRepository repository,
    AppDbContext db,
    IPasswordHasher<User> passwordHasher,
    IDataProtectionProvider dataProtectionProvider,
    IConfiguration configuration) : IUserService
{
    private readonly IDataProtector _resetTokenProtector =
        dataProtectionProvider.CreateProtector("TodoApi.UserPasswordReset.v1");

    private static UserResponse ToResponse(User user) =>
        new(user.Id, user.Username, user.Email, user.IsActive, user.CreatedAt, user.UpdatedAt);

    private async Task<User> GetOrThrowAsync(int id, CancellationToken ct) =>
        await repository.GetByIdAsync(id, ct) ?? throw new KeyNotFoundException($"User {id} not found.");

    public async Task<PaginatedResponse<UserResponse>> GetAllAsync(int page, int pageSize, CancellationToken ct = default)
    {
        page = Math.Max(1, page);
        pageSize = Math.Clamp(pageSize, 1, 100);
        var (users, total) = await repository.GetAllAsync((page - 1) * pageSize, pageSize, ct);
        return new PaginatedResponse<UserResponse>(
            users.Select(ToResponse), total, page, pageSize, (int)Math.Ceiling(total / (double)pageSize));
    }

    public async Task<UserResponse> GetByIdAsync(int id, CancellationToken ct = default) =>
        ToResponse(await GetOrThrowAsync(id, ct));

    public async Task<UserResponse> CreateAsync(CreateUserRequest request, CancellationToken ct = default)
    {
        await EnsureUniqueAsync(request.Username, request.Email, null, ct);
        var user = new User
        {
            Username = request.Username.Trim(),
            Email = NormalizeEmail(request.Email),
            IsActive = request.IsActive,
            CreatedAt = DateTime.UtcNow
        };
        user.PasswordHash = passwordHasher.HashPassword(user, request.Password);
        await repository.AddAsync(user, ct);
        await db.SaveChangesAsync(ct);
        return ToResponse(user);
    }

    public async Task<UserResponse> UpdateAsync(int id, UpdateUserRequest request, CancellationToken ct = default)
    {
        var user = await GetOrThrowAsync(id, ct);
        var username = request.Username?.Trim() ?? user.Username;
        var email = request.Email is null ? user.Email : NormalizeEmail(request.Email);
        await EnsureUniqueAsync(username, email, id, ct);

        user.Username = username;
        user.Email = email;
        if (request.Password is not null)
            user.PasswordHash = passwordHasher.HashPassword(user, request.Password);
        user.UpdatedAt = DateTime.UtcNow;
        repository.Update(user);
        await db.SaveChangesAsync(ct);
        return ToResponse(user);
    }

    public async Task<UserResponse> SetActiveAsync(int id, bool isActive, CancellationToken ct = default)
    {
        var user = await GetOrThrowAsync(id, ct);
        user.IsActive = isActive;
        user.UpdatedAt = DateTime.UtcNow;
        repository.Update(user);
        await db.SaveChangesAsync(ct);
        return ToResponse(user);
    }

    public Task<UserResponse> SignUpAsync(SignUpRequest request, CancellationToken ct = default) =>
        CreateAsync(new CreateUserRequest(request.Username, request.Email, request.Password), ct);

    public Task<UserResponse> GetProfileAsync(int userId, CancellationToken ct = default) => GetByIdAsync(userId, ct);

    public async Task<UserResponse> UpdateProfileAsync(int userId, UpdateProfileRequest request, CancellationToken ct = default)
    {
        var user = await GetOrThrowAsync(userId, ct);
        var username = request.Username?.Trim() ?? user.Username;
        var email = request.Email is null ? user.Email : NormalizeEmail(request.Email);
        await EnsureUniqueAsync(username, email, userId, ct);
        user.Username = username;
        user.Email = email;
        user.UpdatedAt = DateTime.UtcNow;
        repository.Update(user);
        await db.SaveChangesAsync(ct);
        return ToResponse(user);
    }

    public async Task ChangePasswordAsync(int userId, ChangePasswordRequest request, CancellationToken ct = default)
    {
        var user = await GetOrThrowAsync(userId, ct);
        if (!user.IsActive)
            throw new InvalidPasswordException("The user account is inactive.");

        var verification = passwordHasher.VerifyHashedPassword(user, user.PasswordHash, request.CurrentPassword);
        if (verification == PasswordVerificationResult.Failed)
            throw new InvalidPasswordException("The current password is incorrect.");

        user.PasswordHash = passwordHasher.HashPassword(user, request.NewPassword);
        user.UpdatedAt = DateTime.UtcNow;
        repository.Update(user);
        await db.SaveChangesAsync(ct);
    }

    public async Task RequestPasswordResetAsync(ResetPasswordRequest request, CancellationToken ct = default)
    {
        var user = await repository.GetByEmailAsync(NormalizeEmail(request.Email), ct);
        if (user is null || !user.IsActive)
            return; // Do not reveal whether an account exists.

        var lifetimeMinutes = Math.Max(1, configuration.GetValue("PasswordReset:TokenLifetimeMinutes", 60));
        var payload = new PasswordResetTokenPayload(user.Id, DateTime.UtcNow.AddMinutes(lifetimeMinutes), user.PasswordHash);
        var token = _resetTokenProtector.Protect(JsonSerializer.Serialize(payload));
        var baseUrl = configuration["PasswordReset:ConfirmationUrl"] ?? "/reset-password";
        var separator = baseUrl.Contains('?') ? '&' : '?';
        var resetUrl = $"{baseUrl}{separator}token={Uri.EscapeDataString(token)}";

        db.EmailLogs.Add(new EmailLog
        {
            Recipient = user.Email,
            Subject = "Reset your Todo API password",
            Body = $"Use this link to reset your password: {resetUrl}\n\nThis link expires in {lifetimeMinutes} minutes.",
            Status = "pending",
            CreatedAt = DateTime.UtcNow
        });
        await db.SaveChangesAsync(ct);
    }

    public async Task ConfirmPasswordResetAsync(ConfirmPasswordResetRequest request, CancellationToken ct = default)
    {
        PasswordResetTokenPayload payload;
        try
        {
            payload = JsonSerializer.Deserialize<PasswordResetTokenPayload>(_resetTokenProtector.Unprotect(request.Token))
                ?? throw new InvalidPasswordResetTokenException("The password reset token is invalid or expired.");
        }
        catch (InvalidPasswordResetTokenException)
        {
            throw;
        }
        catch (Exception ex) when (ex is System.Security.Cryptography.CryptographicException or JsonException)
        {
            throw new InvalidPasswordResetTokenException("The password reset token is invalid or expired.");
        }

        var user = await repository.GetByIdAsync(payload.UserId, ct);
        if (user is null || !user.IsActive || payload.ExpiresAtUtc < DateTime.UtcNow ||
            !string.Equals(payload.PasswordHash, user.PasswordHash, StringComparison.Ordinal))
            throw new InvalidPasswordResetTokenException("The password reset token is invalid or expired.");

        user.PasswordHash = passwordHasher.HashPassword(user, request.NewPassword);
        user.UpdatedAt = DateTime.UtcNow;
        repository.Update(user);
        await db.SaveChangesAsync(ct);
    }

    private async Task EnsureUniqueAsync(string username, string email, int? excludingId, CancellationToken ct)
    {
        if (await repository.UsernameExistsAsync(username.Trim(), excludingId, ct))
            throw new UserConflictException("Username is already in use.");
        if (await repository.EmailExistsAsync(NormalizeEmail(email), excludingId, ct))
            throw new UserConflictException("Email is already in use.");
    }

    private static string NormalizeEmail(string email) => email.Trim().ToLowerInvariant();

    private sealed record PasswordResetTokenPayload(int UserId, DateTime ExpiresAtUtc, string PasswordHash);
}
